pkg/cloud/services/ec2/bastion.go

```diff
@@ -139,9 +139,16 @@ func (s *Service) getDefaultBastion() *infrav1.Instance {
 	name := fmt.Sprintf("%s-bastion", s.scope.Name())
 	userData, _ := userdata.NewBastion(&userdata.BastionInput{})
 
-	keyName := defaultSSHKeyName
-	if s.scope.AWSCluster.Spec.SSHKeyName != "" {
-		keyName = s.scope.AWSCluster.Spec.SSHKeyName
+	// If SSHKeyName WAS NOT provided (nil) then use the defaultSSHKeyName
+	// If SSHKeyName WAS provided _but is an empty string_, do not set a key pair
+	// Otherwise (SSHKeyName WAS provided and is NOT an empty string), use the provided key pair name
+	var keyName string
+	if s.scope.AWSCluster.Spec.SSHKeyName != nil {
+		if *s.scope.AWSCluster.Spec.SSHKeyName != "" {
+			keyName = *s.scope.AWSCluster.Spec.SSHKeyName
+		}
+	} else {
+		keyName = defaultSSHKeyName
 	}
```

Comment:
I think it might be a bit easier to read if we adjust the logic a bit here.
```suggestion
	keyName := s.scope.AWSCluster.Spec.SSHKeyName
	if keyName == nil {
		keyName = defaultSSHKeyName
	}
```

Reply:
I think that wouldn't handle the case where `SSHKeyName == ""` (which specifies _not_ to use an SSH key).

How about this?
```
	keyName := ""
	if s.scope.AWSCluster.Spec.SSHKeyName == nil {
		keyName = defaultSSHKeyName
	} else {
		keyName = *s.scope.AWSCluster.Spec.SSHKeyName
	}
```

Or should I consider adding a new bool `UseSSHKeys` instead?

Reply:
Oops, I missed something with the type change, the below should work:

```suggestion
	keyName := s.scope.AWSCluster.Spec.SSHKeyName
	if keyName == nil {
		keyName = aws.String(defaultSSHKeyName)
	}
```

The use of keyName below would need to be changed to remove the wrapping `aws.String()` call.

This should work, because if s.scope.AWSCluster.Spec.SSHKeyName == "" (or any non-nil value), then we wouldn't set keyName to a pointer to defaultSSHKeyName, so a user-specified "" would take precedence over the default key as we would expect.

Reply:
This is nice, I've changed the code